### PR TITLE
Fix cases where we miss finalizing tasks

### PR DESCRIPTION
**Pull request by EngHabu**
# TL;DR
_Please replace this text with a description of what this PR accomplishes._

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [] Unit tests added
 - [X] Code documentation added
 - [X] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/lyft/flyte/issues/265
https://github.com/lyft/flyte/issues/318

### CODE REVIEW

File: cmd/controller/cmd/root.go [hunk 1]
```diff
@@ -215,8 +218,18 @@ func executeRootCmd(cfg *config2.Config) {
 		logger.Fatalf(ctx, "Failed to initialize controller run-time manager. Error: %v", err)
 	}
 
-	c, err := controller.New(ctx, cfg, kubeClient, flyteworkflowClient, flyteworkflowInformerFactory, mgr, propellerScope)
+	// Start controller runtime manager to start listening to resource changes.
+	go func(ctx context.Context) {
```
Comment (kumare3) [inline cmd/controller/cmd/root.go:222]: what is this for?

File: pkg/controller/nodes/executor.go [hunk 2]
```diff
@@ -592,12 +592,22 @@ func (c *nodeExecutor) SetInputsForStartNode(ctx context.Context, execContext ex
 	return executors.NodeStatusComplete, nil
 }
 
+func canOperateOnNode(phase v1alpha1.NodePhase) bool {
```
Comment (kumare3) [inline pkg/controller/nodes/executor.go:595]: shouldConsiderNode
shouldTraverseNode
canTraverseNode
canHandleNode?

File: pkg/controller/nodes/handler/transition_info.go [hunk 3]
```diff
@@ -7,6 +7,8 @@ import (
 	"github.com/lyft/flytestdlib/storage"
 )
 
+//go:generate enumer --type=EPhase --trimprefix=EPhase
```
Comment (kumare3) [inline pkg/controller/nodes/handler/transition_info.go:10]: nice

File: pkg/controller/nodes/subworkflow/handler.go [hunk 4]
```diff
@@ -64,26 +65,44 @@ func (w *workflowNodeHandler) Handle(ctx context.Context, nCtx handler.NodeExecu
...
 
 		return invalidWFNodeError()
-	}
+	} else if workflowPhase == v1alpha1.WorkflowNodePhaseExecuting {
+		if wfNode.GetSubWorkflowRef() != nil {
+			return w.subWfHandler.CheckSubWorkflowStatus(ctx, nCtx)
+		} else if wfNode.GetLaunchPlanRefID() != nil {
+			return w.lpHandler.CheckLaunchPlanStatus(ctx, nCtx)
+		}
+	} else if workflowPhase == v1alpha1.WorkflowNodePhaseFailing {
```
Comment (kumare3) [inline pkg/controller/nodes/subworkflow/handler.go:91]: ohh did I already add this phase last time around, thats why I dont see it in this change

**Comment by kumare3** [comment]
This Looks great, tests are failing. I think lets get this one in

File: pkg/controller/nodes/subworkflow/handler.go (hunk 4 above)
Comment (EngHabu) [reply pkg/controller/nodes/subworkflow/handler.go:91]: Yes, it was just not used

File: pkg/apis/flyteworkflow/v1alpha1/iface.go [hunk 5]
```diff
@@ -117,6 +118,8 @@ func (p WorkflowPhase) String() string {
 		return "Succeeding"
 	case WorkflowPhaseAborted:
 		return "Aborted"
+	case WorkflowPhaseHandlingFailureNode:
```
[the remark this answers is not in the document]
Comment (EngHabu) [reply pkg/apis/flyteworkflow/v1alpha1/iface.go:129]: Added a comment [above](https://github.com/lyft/flytepropeller/pull/135/files#diff-ade6cf904f7da4ab15a25e19efc7ca72R102-R110)

Please let me know if it makes sense?
TL;DR; yes it's for handling failure nodes, and yes it's not yet supported in SDK...

I also updated [this](https://github.com/lyft/flytepropeller/pull/135/files#diff-d4cd755eb17448b112df1c4cea0aeafeR70-R81) Which does not do what you wanted fully but it shows a very simple diagram of possible transitions
